File: rudolph/src/main/java/cn/wzbos/android/rudolph/router/FragmentRouter.java
package cn.wzbos.android.rudolph.router;

import android.util.Log;

import cn.wzbos.android.rudolph.exception.ErrorCode;
import cn.wzbos.android.rudolph.exception.ErrorMessage;
import cn.wzbos.android.rudolph.exception.RudolphException;

public class FragmentRouter<R> extends Router<R> {

    FragmentRouter(RouteBuilder<?, ?> builder) {
        super(builder);
    }

    private FragmentRouter(Builder<?, ?> builder) {
        super(builder);
    }

    @Override
    public R open() {
        if (super.intercept(null))
            return null;

        if (target == null) {
            if (callback != null)
                callback.onError(this, new RudolphException(ErrorCode.NOT_FOUND, ErrorMessage.NOT_FOUND_ERROR));
            return null;
        }

        try {
            Object instance = target.getConstructor().newInstance();
            if (instance instanceof android.app.Fragment) {
                ((android.app.Fragment) instance).setArguments(this.bundle);
            } else if (instance instanceof android.support.v4.app.Fragment) {
                ((android.support.v4.app.Fragment) instance).setArguments(this.bundle);
            }

            if (callback != null)
                callback.onSuccess(this);

            return (R) instance;
        } catch (Exception e) {
            if (callback != null) {
                callback.onError(this, new RudolphException(ErrorCode.FRAGMENT_CREATE_FAILED, "Fragment 创建失败！", e));
            } else {
                Log.e("rudolph", "Fragment 创建失败！", e);
            }
        }
        return null;
    }

    public static class Builder<T, R> extends RouteBuilder<Builder<T, R>, FragmentRouter<R>> {

        public Builder(Class<?> cls) {
            super(cls);
        }

        public Builder(String path) {
            super(path);
        }


        @Override
        public FragmentRouter<R> build() {
            return new FragmentRouter<>(this);
        }
    }
}

File: rudolph/src/main/java/cn/wzbos/android/rudolph/router/FragmentRouter.kt
package cn.wzbos.android.rudolph.router

import androidx.fragment.app.Fragment
import cn.wzbos.android.rudolph.exception.ErrorCode
import cn.wzbos.android.rudolph.exception.ErrorMessage
import cn.wzbos.android.rudolph.exception.RudolphException
import cn.wzbos.android.rudolph.logger.RLog

class FragmentRouter<R> : Router<R?> {
    private constructor(builder: Builder<*, *>) : super(builder)
    constructor(builder: UriRouter.Builder<*>) : super(builder)

    @Suppress("UNCHECKED_CAST")
    override fun execute(): R? {
        if (super.intercept(null)) return null
        if (target == null) {
            callback?.onError(
                this,
                RudolphException(ErrorCode.NOT_FOUND, ErrorMessage.NOT_FOUND_ERROR)
            )
            return null
        }

        try {
            val constructor = target?.getConstructor()
            if (constructor == null) {
                RLog.e("FragmentRouter", "$target 必须提供一个无参的构造函数！")
                return null
            }

            val instance = target?.getConstructor()?.newInstance()
            if (instance is Fragment) {
                instance.arguments = extras
            }

            callback?.onSuccess(this)
            return instance as R
        } catch (e: Exception) {
            if (callback != null) {
                callback?.onError(
                    this,
                    RudolphException(ErrorCode.FRAGMENT_CREATE_FAILED, "Fragment 创建失败！", e)
                )
            } else {
                RLog.e("FragmentRouter", "Fragment 创建失败！", e)
            }
        }
        return null
    }

    open class Builder<T, R>(path: String) :
        RouteBuilder<Builder<T, R>?, FragmentRouter<R>?>(path) {

        override fun build(): FragmentRouter<R> {
            return FragmentRouter(this)
        }

        fun execute(): R? {
            return build().execute()
        }
    }
}